calicoctl/commands/import.go

```diff
@@ -0,0 +1,252 @@
+	cf := parsedArgs["--config"].(string)
+	client, err := clientmgr.NewClient(cf)
+	if err != nil {
+		return err
+	}
+
+	// Check that the datastore configured datastore is kubernetes
+	cfg, err := clientmgr.LoadClientConfig(cf)
+	if err != nil {
+		log.Info("Error loading config")
+		return err
+	}
+
+	if cfg.Spec.DatastoreType != apiconfig.Kubernetes {
+		return fmt.Errorf("Invalid datastore type: %s to import to for datastore migration. Datastore type must be kubernetes", cfg.Spec.DatastoreType)
+	}
+
+	// TODO: Attempt to import the CRDs here
+
+	// TODO: On failure, print instructions on how to cleanse datastore
+	/*
+		err = checkCalicoResourcesNotExist(parsedArgs)
+		if err != nil {
+			return fmt.Errorf("Datastore already has Calico resources: %s. INSERT INSTRUCTIONS TO DELETE", err)
+		}
+	*/
+
+	// Ensure that the cluster info resource is initialized.
+	ctx := context.Background()
+	if err := client.EnsureInitialized(ctx, "", ""); err != nil {
+		return fmt.Errorf("Unable to initialize cluster information for the datastore migration: %s", err)
+	}
+	// TODO: Need to ensure the datastore is locked here.
+
+	// Split file into v3 API, ClusterGUID, and IPAM components
+	filename := parsedArgs["--filename"].(string)
+	v3Yaml, clusterInfoJson, ipamJson, err := splitImportFile(filename)
+	if err != nil {
+		return fmt.Errorf("Error while reading migration file: %s\n", err)
+	}
+
+	// Apply v3 API resources
+	err = updateV3Resources(v3Yaml)
+	if err != nil {
+		return fmt.Errorf("Failed to import v3 resources: %s\n", err)
+	}
+
+	// Create the cluster info resource. This should not exist since calico-node was only writing to the etcd datastore.
+	err = updateClusterInfo(ctx, client, clusterInfoJson)
+	if err != nil {
+		return fmt.Errorf("Failed to update cluster information: %s", err)
+	}
+
+	// Import IPAM components
+	fmt.Print("Importing IPAM resources\n")
+	ipam := NewMigrateIPAM(client)
+	err = json.Unmarshal(ipamJson, ipam)
+	if err != nil {
+		return fmt.Errorf("Failed to read IPAM resources: %s\n", err)
+	}
+	results := ipam.PushToDatastore()
+
+	// Handle the IPAM results
+	if results.numHandled == 0 {
+		if results.numResources == 0 {
+			return fmt.Errorf("No IPAM resources specified in file")
+		} else {
+			return fmt.Errorf("Failed to import any IPAM resources: %v", results.resErrs)
+		}
+	} else if len(results.resErrs) == 0 {
+		fmt.Printf("Successfully applied %d IPAM resource(s)\n", results.numHandled)
+	} else {
+		if results.numHandled != 0 && len(results.resErrs) > 0 {
+			fmt.Printf("Partial success: ")
+			fmt.Printf("applied the first %d out of %d resources:\n", results.numHandled, results.numResources)
+		}
+		return fmt.Errorf("Hit error(s): %v", results.resErrs)
+	}
+
+	fmt.Print("Datastore information successfully imported. To complete the datastore migration, run `calicoctl unlock` and modify your calico configuration to match the kubenretes datastore.\n")
```

Comment:
I think this instruction is wrong - users will need to do a rolling update to point calico at kdd before running unlock.